accounting/src/creditcommons/credit-commons-controller.ts

```diff
@@ -35,9 +55,72 @@ export class CreditCommonsControllerImpl extends AbstractCurrencyController impl
     if (record.lastHash !== ctx.lastHashAuth?.lastHash) {
       throw unauthorized(`value of last-hash header ${JSON.stringify(ctx.lastHashAuth?.lastHash)} does not match our records.`)
     }
+    return record.vostroId
   }
   async getWelcome(ctx: Context) {
     await this.checkLastHashAuth(ctx)
     return { message: 'Welcome to the Credit Commons federation protocol.' }
   }
+  async getAccountHistory(ctx: Context) {
+    await this.checkLastHashAuth(ctx)
+    return {
+      data: {
+        '2025-04-01 09:22:37': 0,
+        '2025-04-01 09:22:41': 0,
+        '2025-04-01 09:22:42': -60
+      },
+      meta: {
+        min: -60,
+        max: 0,
+        points: 3,
+        start: '2025-04-01 09:22:37',
+        end: '2025-04-01 09:22:42'
+      }
+    };
+  }
+  async createTransaction(ctx: Context, transaction: CreditCommonsTransaction) {
+    this.gatewayAccountId = await this.checkLastHashAuth(ctx)
+    
+    let localTransfers: InputTransfer[] = []
+    let netGain = 0;
+    let recipient = null;
+    for (let i=0; i < transaction.entries.length; i++) {
+      let payer, payee, thisRecipient;
+      if (transaction.entries[i].payer.startsWith(this.ledgerBase)) {
+        payer = transaction.entries[i].payer.slice(this.ledgerBase.length)
+        payee = this.gatewayAccountId
+        thisRecipient = payer
+        netGain -= transaction.entries[i].quant
+      }
+      if (transaction.entries[i].payee.startsWith(this.ledgerBase)) {
+        payee = transaction.entries[i].payee.slice(this.ledgerBase.length)
+        payer = this.gatewayAccountId
+        if (thisRecipient) {
+          throw new Error('Payer and Payee cannot both be local')
+        }
+        thisRecipient = payee
+        netGain += transaction.entries[i].quant
+      }
+      if (!thisRecipient) {
+        throw new Error('Payer and Payee cannot both be remote')
+      }
+      if (recipient && recipient !== thisRecipient) {
+        throw new Error('All entries must be to or from the same local account')
+      }
+      recipient = thisRecipient
+      localTransfers.push({
+        id: `${transaction.uuid}-${i}`,
+        state: 'committed',
+        amount: transaction.entries[i].quant * this.ledgerScale,
+        meta: transaction.entries[i].description,
+        payer: { id: payer!, type: 'account' },
+        payee: { id: payee!, type: 'account' },
+      })
+    }
+    if (netGain <= 0) {
+      throw new Error('Net gain must be positive')
+    }
+    await this.transferController.createMultipleTransfers(systemContext(), localTransfers)
```

Comment:
Maybe we could add a comment here saying that we'd like to use a function so that all transfers either succeed or fail. I've created a GH issue for that:
https://github.com/komunitin/komunitin/issues/368

Reply:
Right! For now, as discussed in Telegram, I changed it so they all get shmushed into one transfer (amounts added up and descriptions concatenated into one long `meta` string). The reason for this was mainly that I need to store the CC transfer id and this way I can store it in the `Transfers` table and don't need to create a separate `TransferCollections` table.